### DISCUSSION TITLE
Bug with ManyToMany relation save

### DISCUSSION THREAD

**Discussion by iam-mishgun** [mikro-orm/mikro-orm#4387] [Q&A] [answered]
Hi!
I have strange issue. I want to create 'Space' with categories (which already created). But when I try save new 'Space' I got a message that shows trying to create a new 'Category' rather than using an existing one (and crash report because id already exists).
How I can create 'Space' with already created 'Categories'?
Error:
```sql
insert into \"category\" (\"id\", \"created_at\", \"updated_at\", \"name\") values ('5b5b2663-8d4d-4311-a2f7-39935a496a92', '2023-05-25T10:11:11.890Z', '2023-05-25T10:11:11.890Z', '') returning \"id\", \"created_at\", \"updated_at\", \"name\" - duplicate key value violates unique constraint \"category_pkey\"
```

Entities:
```typescript
@Entity()
export class Category extends BaseEntity {
  @PrimaryKey({ columnType: 'uuid', defaultRaw: 'gen_random_uuid()' })
  id: string = v4();

  @Property({ columnType: 'text' })
  name: string;

  @ManyToMany(() => Space)
  spaces = new Collection<Space>(this);
}
@Entity()
export class Space extends BaseEntity {
  @PrimaryKey({ columnType: 'text' })
  id!: string;

  @Property({ columnType: 'text' })
  name: string;

  @ManyToMany(() => Category, 'spaces')
  categories = new Collection<Category>(this);
}
```
Service:
```ts
export class CreateSpaceInput {
  @IsString()
  @MinLength(1)
  @MaxLength(255)
  @IsUniqueSpaceId()
  @Transform(({ value }) => value.toLowerCase())
  id: string;

  categories: {id: string}[];
}

@Injectable()
export class SpaceService {
  constructor(
    @InjectRepository(Space) private readonly spaceRepository: EntityRepository<Space>,
  ) {}
  public async create(data: CreateSpaceInput): Promise<Space | null> {
    const space = this.spaceRepository.assign(new Space(), spaceToCreate);
    this.spaceRepository.getEntityManager().persistAndFlush(space);
    return space;
  }
}
```

**Comment by B4nan** [comment, marked as answer]
> `categories: {id: string}[];`

Instead of passing in objects with `id` property, pass in just the value of the FK - then it will be considered as an existing entity instead of a new one.

Few general notes:
- you shouldn't really do these shenanigans with the repository, just import `EntityManager` and work with that
- instead of creating the instance via the constructor and using `assign`, you can use `em.create` to do both in one step

**Reply by superiums** [reply]
Maybe it's better to accept objects with `id` property, 
may we make mikro to decide weather to create a new relation just by cascade config ?

**Reply by B4nan** [reply]
That wouldn't work with client-side generated PK, e.g. with UUIDs. This has nothing to do with the `cascade` option.

**Comment by iam-mishgun** [comment]
Thank you!

**Comment by iam-mishgun** [comment]
@B4nan having strange issue when sending `categories: string[]`
```log
[16585] LOG:  statement: BEGIN;
[16585] LOG:  statement: insert into "space" ("id", "created_at", "updated_at", "name", "brief", "logo", "cover", "is_active") values ('trololo', '2023-05-25T12:35:06.696Z', '2023-05-25T12:35:06.696Z', 'tralalal', 'asdasdasdas asd', '123', '321', true) returning "id", "created_at", "updated_at", "is_active"
[16585] LOG:  statement: COMMIT;
```
They dont try create relation

**Reply by B4nan** [reply]
Hmm right, because `categories` is an inverse side, you should be working with the owning side, or with initialized values instead. I really need to find a way around this in v6, or at least add a validation, this is really confusing.

I'd say it makes more sense to flip the sides, make `categories` owning side instead, it feels semantically more correct - as it is not the category that owns spaces (which is what you have right now). 

FYI owning side is either the one that specifies `owner: true` in the options, or the one that is not using the `mappedBy` option/parameter.

**Reply by iam-mishgun** [reply]
```ts
// on space side
  @ManyToMany({ entity: () => Category, owner: true })
  categories: Collection<Category> | string[] = new Collection<Category>(this);
// on category side 
  @ManyToMany({ entity: () => Space, mappedBy: 'categories' })
  spaces = new Collection<Space>(this);
  ```
  still not working

**Reply by B4nan** [reply]
I need to see more than that, ideally provide complete reproduction.

> categories: Collection<Category> | string[]

Note that this will be always a `Collection`, the `string[]` part is wrong, and you should never put there just an array of strings - maybe you are doing that?

**Reply by iam-mishgun** [reply]
`string[]` for request side
Etities:
```ts
@Entity()
export class Category extends BaseEntity {
  @PrimaryKey({ columnType: 'uuid', defaultRaw: 'gen_random_uuid()' })
  id: string = v4();

  @Property({ columnType: 'text' })
  name: string;

  @ManyToMany({ entity: () => Space, mappedBy: 'categories' })
  spaces = new Collection<Space>(this);
}

@Entity()
export class Space extends BaseEntity {
  @ManyToMany({ entity: () => Category, owner: true })
  categories: Collection<Category> | string[] = new Collection<Category>(this);
}
```
`CreateSpaceInput` looks like 
```ts
{
categories: string[]
}
```
Other not changed

**Reply by B4nan** [reply]
There is no request side of entity, the entity is not a DTO and vice versa, those are two things. `Collection<Category> | string[]` is wrong.

Also, I said I want to see the code, you gave me entity definition again, the problem is how you work with the entities.

**Reply by iam-mishgun** [reply]
Request: 
```json
{
  "id": "trololo",
  "name": "tralalal",
  "brief": "asdasdasdas asd",
  "logo": "123",
  "cover": "321",
  "isActive": true,
  "categories": [ "bc6d2701-778f-4de2-80d8-7175969fcd70" ]
}
```
Service method:
```
public async create(data: CreateSpaceInput): Promise<Space | null> {
    const space = this.spaceRepository.create(data as RequiredEntityData<Space>);
    await this.spaceRepository.getEntityManager().persistAndFlush(space);
    return space;
 }
 ```

**Reply by iam-mishgun** [reply]
"@mikro-orm/core": "^5.7.10",
